File: Sources/ClaudePet/UpdateChecker.swift
import Cocoa

class UpdateChecker {
    let currentVersion: String = {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.0"
    }()
    private let repoAPI = "https://api.github.com/repos/devTheKiwi/ClaudePet/releases/latest"

    var latestVersion: String?
    var updateAvailable: Bool = false
    var onResult: ((String) -> Void)?

    func checkOnLaunch() {
        DispatchQueue.global().asyncAfter(deadline: .now() + 5) { [weak self] in
            self?.check(announceWhenCurrent: true)
        }
    }

    private var lastCheckAt: Date?

    /// 주기적/이벤트 재확인 — 새 버전일 때만 알림. 30분 쿨다운(깨어남·활성화 연타로 API 도배 방지)
    func checkPeriodic() {
        if let last = lastCheckAt, Date().timeIntervalSince(last) < 1800 { return }
        lastCheckAt = Date()
        DispatchQueue.global().async { [weak self] in
            self?.check(announceWhenCurrent: false)
        }
    }

    private func check(announceWhenCurrent: Bool) {
        guard let url = URL(string: repoAPI) else {
            if announceWhenCurrent {
                DispatchQueue.main.async { self.onResult?(L10n.isKorean ? "업데이트 확인 실패" : "Update check failed") }
            }
            return
        }

        let task = URLSession.shared.dataTask(with: url) { [weak self] data, _, error in
            guard let self = self else { return }

            guard error == nil,
                  let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let latestTag = json["tag_name"] as? String else {
                if announceWhenCurrent {
                    DispatchQueue.main.async { self.onResult?(L10n.isKorean ? "업데이트 확인 실패" : "Update check failed") }
                }
                return
            }

            let latest = latestTag.replacingOccurrences(of: "v", with: "")
            self.latestVersion = latest

            if self.isNewer(latest: latest, current: self.currentVersion) {
                self.updateAvailable = true
                DispatchQueue.main.async {
                    self.onResult?(L10n.updateAvailable(latest))
                }
            } else {
                self.updateAvailable = false
                if announceWhenCurrent {
                    DispatchQueue.main.async {
                        self.onResult?(L10n.updateLatest(self.currentVersion))
                    }
                }
            }
        }
        task.resume()
    }

    private func isNewer(latest: String, current: String) -> Bool {
        let l = latest.split(separator: ".").compactMap { Int($0) }
        let c = current.split(separator: ".").compactMap { Int($0) }
        // 현재 버전을 못 읽으면(빈 값/파싱 실패) 업데이트 알림을 띄우지 않음 — 무한 알림 방지
        if l.isEmpty || c.isEmpty { return false }
        for i in 0..<max(l.count, c.count) {
            let lv = i < l.count ? l[i] : 0
            let cv = i < c.count ? c[i] : 0
            if lv > cv { return true }
            if lv < cv { return false }
        }
        return false
    }

    func runUpdate() {
        let installCommand = "curl -sL https://raw.githubusercontent.com/devTheKiwi/ClaudePet/main/remote-install.sh | bash"

        // .command 임시 파일로 터미널 실행 (권한 문제 없음)
        let tmpFile = "/tmp/claudepet-update.command"
        let script = "#!/bin/bash\n\(installCommand)\nrm -f \(tmpFile)\n"
        do {
            try script.write(toFile: tmpFile, atomically: true, encoding: .utf8)
            try FileManager.default.setAttributes([.posixPermissions: 0o755], ofItemAtPath: tmpFile)
            NSWorkspace.shared.open(URL(fileURLWithPath: tmpFile))
        } catch {
            // 실패 시 클립보드에 복사
            NSPasteboard.general.clearContents()
            NSPasteboard.general.setString(installCommand, forType: .string)
        }
    }
}

File: windows/src/main/update-checker.ts
/**
 * UpdateChecker.swift 포팅
 *
 * GitHub Releases API 로 새 버전 체크.
 * Windows 에선 자동 업데이트 대신 GitHub Releases 페이지로 이동.
 */

import { shell } from 'electron';
import * as S from './strings';

const REPO_API = 'https://api.github.com/repos/devTheKiwi/ClaudePet/releases/latest';
const RELEASES_PAGE = 'https://github.com/devTheKiwi/ClaudePet/releases/latest';

export class UpdateChecker {
  public readonly currentVersion = '2.7.0';
  public latestVersion: string | null = null;
  public updateAvailable = false;
  public onResult: ((message: string) => void) | null = null;

  checkOnLaunch(): void {
    setTimeout(() => this.check(true), 5000);
  }

  private lastCheckAt = 0;

  /** 주기적/이벤트 재확인 — 새 버전일 때만 알림. 30분 쿨다운(깨어남·포커스 연타로 API 도배 방지) */
  checkPeriodic(): void {
    const now = Date.now();
    if (now - this.lastCheckAt < 30 * 60 * 1000) return;
    this.lastCheckAt = now;
    this.check(false);
  }

  private async check(announceWhenCurrent: boolean): Promise<void> {
    try {
      const res = await fetch(REPO_API, {
        headers: { 'User-Agent': 'ClaudePet-Windows' },
      });
      if (!res.ok) {
        if (announceWhenCurrent) this.onResult?.(S.updateFailed);
        return;
      }
      const json = (await res.json()) as { tag_name?: string };
      const tag = json.tag_name;
      if (!tag) {
        if (announceWhenCurrent) this.onResult?.(S.updateFailed);
        return;
      }
      const latest = tag.replace(/^v/, '');
      this.latestVersion = latest;

      if (this.isNewer(latest, this.currentVersion)) {
        this.updateAvailable = true;
        this.onResult?.(S.updateAvailable(latest));
      } else {
        this.updateAvailable = false;
        if (announceWhenCurrent) this.onResult?.(S.updateLatest(this.currentVersion));
      }
    } catch {
      if (announceWhenCurrent) this.onResult?.(S.updateFailed);
    }
  }

  private isNewer(latest: string, current: string): boolean {
    const l = latest.split('.').map((s) => parseInt(s, 10) || 0);
    const c = current.split('.').map((s) => parseInt(s, 10) || 0);
    const len = Math.max(l.length, c.length);
    for (let i = 0; i < len; i++) {
      const lv = l[i] ?? 0;
      const cv = c[i] ?? 0;
      if (lv > cv) return true;
      if (lv < cv) return false;
    }
    return false;
  }

  /** Releases 페이지 열기 */
  runUpdate(): void {
    shell.openExternal(RELEASES_PAGE);
  }
}
